File: web/lib/csf.ts
/**
 * CSF 2.0 API Library
 * 
 * Fast, memoized access to NIST CSF 2.0 taxonomy data
 * with error handling and loading states for grid interfaces.
 */

import { apiFetch } from './api';
import type { 
  CSFTaxonomyResponse, 
  CSFFunction, 
  CSFCategory, 
  CSFSubcategory 
} from '../types/csf';

// Memoization cache
const csfCache = new Map<string, any>();
const CACHE_TTL = 300000; // 5 minutes in milliseconds

interface CacheItem<T> {
  data: T;
  timestamp: number;
}

/**
 * Get cached item if valid, otherwise return null
 */
function getCachedItem<T>(key: string): T | null {
  const item = csfCache.get(key) as CacheItem<T> | undefined;
  if (!item) return null;
  
  const isExpired = Date.now() - item.timestamp > CACHE_TTL;
  if (isExpired) {
    csfCache.delete(key);
    return null;
  }
  
  return item.data;
}

/**
 * Set cache item with timestamp
 */
function setCacheItem<T>(key: string, data: T): void {
  csfCache.set(key, {
    data,
    timestamp: Date.now()
  });
}

/**
 * Get complete CSF 2.0 taxonomy with memoization
 * 
 * Returns full hierarchical structure optimized for grid rendering.
 * Cached for 5 minutes to meet p95 < 2s performance target.
 */
export async function getFunctions(): Promise<CSFFunction[]> {
  const cacheKey = 'csf:functions';
  
  // Check cache first
  const cached = getCachedItem<CSFFunction[]>(cacheKey);
  if (cached) {
    return cached;
  }
  
  try {
    const response: CSFTaxonomyResponse = await apiFetch('/api/v1/csf/functions', {
      timeout: 10000 // 10s timeout for initial load
    });
    
    const functions = response.functions;
    
    // Cache the result
    setCacheItem(cacheKey, functions);
    
    return functions;
  } catch (error) {
    console.error('Error loading CSF functions:', error);
    throw new Error(
      error instanceof Error 
        ? error.message 
        : 'Failed to load CSF taxonomy'
    );
  }
}

/**
 * Get specific CSF function by ID
 */
export async function getFunctionById(functionId: string): Promise<CSFFunction | null> {
  const cacheKey = `csf:function:${functionId}`;
  
  // Check cache first
  const cached = getCachedItem<CSFFunction>(cacheKey);
  if (cached) {
    return cached;
  }
  
  try {
    const function_data: CSFFunction = await apiFetch(`/api/v1/csf/functions/${functionId}`, {
      timeout: 5000
    });
    
    // Cache the result
    setCacheItem(cacheKey, function_data);
    
    return function_data;
  } catch (error) {
    if (error instanceof Error && error.message.includes('404')) {
      return null;
    }
    console.error(`Error loading CSF function ${functionId}:`, error);
    throw new Error(
      error instanceof Error 
        ? error.message 
        : `Failed to load CSF function ${functionId}`
    );
  }
}

/**
 * Get categories, optionally filtered by function
 */
export async function getCategories(functionId?: string): Promise<CSFCategory[]> {
  const cacheKey = functionId ? `csf:categories:${functionId}` : 'csf:categories';
  
  // Check cache first
  const cached = getCachedItem<CSFCategory[]>(cacheKey);
  if (cached) {
    return cached;
  }
  
  try {
    const url = functionId 
      ? `/api/v1/csf/categories?function_id=${functionId}`
      : '/api/v1/csf/categories';
      
    const categories: CSFCategory[] = await apiFetch(url, {
      timeout: 5000
    });
    
    // Cache the result
    setCacheItem(cacheKey, categories);
    
    return categories;
  } catch (error) {
    console.error('Error loading CSF categories:', error);
    throw new Error(
      error instanceof Error 
        ? error.message 
        : 'Failed to load CSF categories'
    );
  }
}

/**
 * Get subcategories, optionally filtered by function and/or category
 */
export async function getSubcategories(
  functionId?: string, 
  categoryId?: string
): Promise<CSFSubcategory[]> {
  const cacheKey = `csf:subcategories:${functionId || 'all'}:${categoryId || 'all'}`;
  
  // Check cache first
  const cached = getCachedItem<CSFSubcategory[]>(cacheKey);
  if (cached) {
    return cached;
  }
  
  try {
    const params = new URLSearchParams();
    if (functionId) params.append('function_id', functionId);
    if (categoryId) params.append('category_id', categoryId);
    
    const url = `/api/v1/csf/subcategories${params.toString() ? '?' + params.toString() : ''}`;
    
    const subcategories: CSFSubcategory[] = await apiFetch(url, {
      timeout: 5000
    });
    
    // Cache the result
    setCacheItem(cacheKey, subcategories);
    
    return subcategories;
  } catch (error) {
    console.error('Error loading CSF subcategories:', error);
    throw new Error(
      error instanceof Error 
        ? error.message 
        : 'Failed to load CSF subcategories'
    );
  }
}

/**
 * Clear CSF cache (useful for testing or forced refresh)
 */
export function clearCSFCache(): void {
  const keys = Array.from(csfCache.keys());
  keys.forEach(key => {
    if (key.startsWith('csf:')) {
      csfCache.delete(key);
    }
  });
}

/**
 * Get cache statistics (for debugging)
 */
export function getCSFCacheStats() {
  const csfKeys = Array.from(csfCache.keys()).filter(key => key.startsWith('csf:'));
  return {
    totalItems: csfKeys.length,
    keys: csfKeys,
    cacheSize: csfCache.size
  };
}

File: app/api/routes/csf.py
"""
CSF 2.0 Grid API

Provides fast access to NIST Cybersecurity Framework 2.0 taxonomy data
including Functions, Categories, and Subcategories for assessment grid structures.
"""

import logging
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Depends, Request
from pydantic import BaseModel

from domain.models import CSFFunction, CSFCategory, CSFSubcategory
from services.csf_taxonomy import get_csf_service
import logging


router = APIRouter(prefix="/api/v1/csf", tags=["csf"])
logger = logging.getLogger(__name__)


class CSFTaxonomyResponse(BaseModel):
    """Full CSF 2.0 taxonomy response"""
    version: str
    functions: List[CSFFunction]
    metadata: dict


class CSFErrorResponse(BaseModel):
    """Error response for CSF endpoints"""
    error: str
    details: Optional[str] = None


def get_correlation_id(request: Request) -> str:
    """Extract correlation ID from request headers"""
    return request.headers.get("X-Correlation-ID", "csf-request")


@router.get("/functions", response_model=CSFTaxonomyResponse)
async def get_csf_functions(request: Request):
    """
    Get complete CSF 2.0 taxonomy with Functions → Categories → Subcategories.
    
    Returns the full hierarchical structure of NIST CSF 2.0 framework
    optimized for grid-based assessment interfaces.
    
    Response is cached for performance with p95 < 2s target.
    """
    correlation_id = get_correlation_id(request)
    
    try:
        logger.info(
            "CSF functions endpoint accessed",
            extra={"correlation_id": correlation_id}
        )
        
        # Get CSF service and load taxonomy
        csf_service = get_csf_service()
        functions = csf_service.get_functions()
        
        # Get version and metadata from raw taxonomy
        taxonomy_data = csf_service.load_csf_taxonomy()
        
        response = CSFTaxonomyResponse(
            version=taxonomy_data.get("version", "2.0"),
            functions=functions,
            metadata=taxonomy_data.get("metadata", {})
        )
        
        logger.info(
            "CSF taxonomy loaded successfully",
            extra={
                "correlation_id": correlation_id,
                "functions_count": len(functions),
                "categories_count": sum(len(f.categories) for f in functions),
                "subcategories_count": sum(
                    len(c.subcategories) for f in functions for c in f.categories
                )
            }
        )
        
        return response
        
    except FileNotFoundError as e:
        logger.error(
            "CSF taxonomy file not found",
            extra={
                "correlation_id": correlation_id,
                "error": str(e)
            }
        )
        raise HTTPException(
            status_code=500,
            detail="CSF taxonomy data not available"
        )
        
    except ValueError as e:
        logger.error(
            "Invalid CSF taxonomy data",
            extra={
                "correlation_id": correlation_id,
                "error": str(e)
            }
        )
        raise HTTPException(
            status_code=500,
            detail="Invalid CSF taxonomy format"
        )
        
    except Exception as e:
        logger.error(
            "Unexpected error loading CSF taxonomy",
            extra={
                "correlation_id": correlation_id,
                "error": str(e)
            }
        )
        raise HTTPException(
            status_code=500,
            detail="Failed to load CSF taxonomy"
        )


@router.get("/functions/{function_id}", response_model=CSFFunction)
async def get_csf_function_by_id(function_id: str, request: Request):
    """
    Get a specific CSF function by ID with its categories and subcategories.
    
    Args:
        function_id: CSF function identifier (e.g., "GV", "ID", "PR", "DE", "RS", "RC")
    
    Returns:
        Complete function with nested categories and subcategories
    """
    correlation_id = get_correlation_id(request)
    
    try:
        logger.info(
            "CSF function by ID endpoint accessed",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id
            }
        )
        
        csf_service = get_csf_service()
        function = csf_service.get_function_by_id(function_id)
        
        if function is None:
            logger.warning(
                "CSF function not found",
                extra={
                    "correlation_id": correlation_id,
                    "function_id": function_id
                }
            )
            raise HTTPException(
                status_code=404,
                detail=f"CSF function '{function_id}' not found"
            )
        
        logger.info(
            "CSF function retrieved successfully",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id,
                "categories_count": len(function.categories),
                "subcategories_count": sum(len(c.subcategories) for c in function.categories)
            }
        )
        
        return function
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(
            "Error retrieving CSF function",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id,
                "error": str(e)
            }
        )
        raise HTTPException(
            status_code=500,
            detail=f"Failed to retrieve CSF function '{function_id}'"
        )


@router.get("/categories", response_model=List[CSFCategory])
async def get_csf_categories(
    function_id: Optional[str] = None, 
    request: Request = None
):
    """
    Get CSF categories, optionally filtered by function.
    
    Args:
        function_id: Optional function ID to filter categories
    
    Returns:
        List of categories with their subcategories
    """
    correlation_id = get_correlation_id(request)
    
    try:
        logger.info(
            "CSF categories endpoint accessed",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id
            }
        )
        
        csf_service = get_csf_service()
        categories = csf_service.get_categories(function_id)
        
        logger.info(
            "CSF categories retrieved successfully",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id,
                "categories_count": len(categories),
                "subcategories_count": sum(len(c.subcategories) for c in categories)
            }
        )
        
        return categories
        
    except Exception as e:
        logger.error(
            "Error retrieving CSF categories",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id,
                "error": str(e)
            }
        )
        raise HTTPException(
            status_code=500,
            detail="Failed to retrieve CSF categories"
        )


@router.get("/subcategories", response_model=List[CSFSubcategory])
async def get_csf_subcategories(
    function_id: Optional[str] = None,
    category_id: Optional[str] = None,
    request: Request = None
):
    """
    Get CSF subcategories, optionally filtered by function and/or category.
    
    Args:
        function_id: Optional function ID to filter subcategories
        category_id: Optional category ID to filter subcategories
    
    Returns:
        List of subcategories
    """
    correlation_id = get_correlation_id(request)
    
    try:
        logger.info(
            "CSF subcategories endpoint accessed",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id,
                "category_id": category_id
            }
        )
        
        csf_service = get_csf_service()
        subcategories = csf_service.get_subcategories(function_id, category_id)
        
        logger.info(
            "CSF subcategories retrieved successfully",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id,
                "category_id": category_id,
                "subcategories_count": len(subcategories)
            }
        )
        
        return subcategories
        
    except Exception as e:
        logger.error(
            "Error retrieving CSF subcategories",
            extra={
                "correlation_id": correlation_id,
                "function_id": function_id,
                "category_id": category_id,
                "error": str(e)
            }
        )
        raise HTTPException(
            status_code=500,
            detail="Failed to retrieve CSF subcategories"
        )